### PR TITLE
Add `IntPtr` conversion APIs for some runtime handles

### LINKED ISSUE #67090: [API Proposal]: Add support for converting to/from `IntPtr` for some handles
### Background and motivation

The `Type.GetTypeFromHandleUnsafe` method allows you to retrieve the `Type` object associated with a type handle pointer retrieved via `RuntimeTypeHandle.Value`.

Currently, this API is `internal` with no `public`-facing equivalent. This means there is currently no safe way for you to get the original `Type` or `RuntimeTypeHandle` value back, leading to the `RuntimeTypeHandle.Value` property being of little use.

Having a way to retrieve the original `Type` or `RuntimeTypeHandle` would also allow the type handle pointer to be used to easily pass types to native code (such as when writing a custom runtime host).

### API Proposal

```diff
public struct RuntimeTypeHandle
{
+     public static RuntimeTypeHandle FromIntPtr(IntPtr value);
+     public static IntPtr ToIntPtr(RuntimeTypeHandle value);
}

public struct RuntimeFieldHandle
{
+    public static RuntimeFieldHandle FromIntPtr(IntPtr value);
+    public static IntPtr ToIntPtr(RuntimeFieldHandle value);
}

public struct RuntimeMethodHandle
{
+    public static RuntimeMethodHandle FromIntPtr(IntPtr value);
+    public static IntPtr ToIntPtr(RuntimeMethodHandle value);
}
```

### API Usage

```C#
var api = new NativeAPI
{
    GetTypeHandle      = &GetTypeHandle,
    GetFullTypeName    = &GetFullTypeName,
    GetFunctionPointer = &GetFunctionPointer
};

// This native method may be interested in querying for
// pointers to UnmanagedCallersOnly methods to invoke.
SomeNativeMethod(&api);
```
```C#
[UnmanagedCallersOnly]
static IntPtr GetTypeHandle(IntPtr typeName)
{
    return Type.GetType(Marshal.PtrToStringUni(typeName)).TypeHandle.Value;
}

[UnmanagedCallersOnly]
static IntPtr GetFunctionPointer(IntPtr type, IntPtr methodName)
{
    var type   = Type.GetTypeFromHandle(RuntimeTypeHandle.FromIntPtr(type));
    var method = type.GetMethod(Marshal.PtrToStringUni(methodName), BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic);
    return method.MethodHandle.GetFunctionPointer();
}

[UnmanagedCallersOnly]
static IntPtr GetFullTypeName(IntPtr type)
{
    return Marshal.StringToHGlobalUni(Type.GetTypeFromHandle(RuntimeTypeHandle.FromIntPtr(type))?.FullName);
}
```

### Alternative Designs

_No response_

### Risks

Being an unmanaged pointer, `RuntimeTypeHandle.Value` will not keep a collectible type alive. This would need to be clearly called out in documentation and could be a potential problem when the API is not used correctly.

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Fixes #67090

/cc @DaZombieKiller

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 1]
```diff
@@ -68,6 +68,34 @@ internal static Type GetTypeHelper(Type typeStart, Type[]? genericArgs, IntPtr p
...
+        /// <param name="value">A <see cref="RuntimeTypeHandle"/> object to retrieve an internal pointer representation from.</param>
+        /// <returns>An <see cref="IntPtr"/> object that represents a <see cref="RuntimeTypeHandle"/> object.</returns>
+        public static IntPtr ToIntPtr(RuntimeTypeHandle value) => value.Value;
+
+        /// <summary>
+        /// A new <see cref="RuntimeTypeHandle"/> object created from a handle to a RuntimeType.
+        /// </summary>
+        /// <param name="value">An IntPtr handle to a RuntimeType to create a <see cref="RuntimeTypeHandle"/> object from.</param>
+        /// <returns>A new <see cref="RuntimeTypeHandle"/> object that corresponds to the value parameter.</returns>
+        public static explicit operator RuntimeTypeHandle(IntPtr value) => FromIntPtr(value);
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:90]: I am not sure we need the operators that are redundant with the methods. This is very advanced scenario that does not need convenience.

I had people ask me more than once about whether it is better to use the IntPtr operators or the explicit methods on GCHandle. I always tell them to use the explicit methods: The explicit methods make the intent more obvious.

**Comment by jkotas** [comment]
I think this will also need an implementation for NativeAOT.

**Comment by jkotas** [comment]
cc @joncham @joshpeterson These methods are useful for implementing Mono embedding API equivalent.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 2]
```diff
@@ -823,6 +851,15 @@ RuntimeMethodHandleInternal Value
         {
             get;
         }
+
+        internal static IRuntimeMethodInfo Create(IntPtr ptr) => new IRuntimeMethodInfoImp(ptr);
+
+        private sealed class IRuntimeMethodInfoImp : IRuntimeMethodInfo
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:857]: I do not think that this implementation is going to work. Some place in reflection assume that types implementing have certain fields at certain offsets - check `RuntimeMethodInfoStub`. This needs to use `RuntimeMethodInfoStub` or something very similar.

(Also, add test that will hit this bug.)

Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:843]: Hmmm. I need to look into this more. I'm not entirely sure how this works and where these assumptions are made.

**Review by jkotas — APPROVED** [review]

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 3]
```diff
@@ -1143,6 +1176,9 @@ RuntimeFieldHandleInternal Value
     [StructLayout(LayoutKind.Sequential)]
     internal sealed class RuntimeFieldInfoStub : IRuntimeFieldInfo
     {
+        public RuntimeFieldInfoStub(IntPtr ptr)
+            => m_fieldHandle = new RuntimeFieldHandleInternal(ptr);
+
         // These unused variables are used to ensure that this class has the same layout as RuntimeFieldInfo
 #pragma warning disable 414, 169
         private object? m_keepalive;
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:1184]: Does this need to initialize KeepAlive, same as `RuntimeMethodHandle` ?

Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:1184]: That is a good question. This wasn't done previously so I don't think so, but it is possible this has a bug in its use. I'm not entirely sure the reason for the keep alive logic so if you can share or perhaps point me to where an example of need is that would be helpful.

Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:1184]: The keep alive logic is for collectible assemblies. It is there to ensure that the RuntimeFieldHandle will prevent the assembly from getting unloaded.

Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:1184]: > This wasn't done previously so I don't think so

The only is place where I see RuntimeFieldInfoStub created is here: https://github.com/dotnet/runtime/blob/6d21c938b2d420b9e06107e409aec7a01b376f59/src/coreclr/vm/field.cpp#L814-L815 . It is setting keeplive correctly.

**Pull request merged**